core/src/main/java/org/akaza/openclinica/service/rule/RuleSetService.java

```diff
@@ -849,29 +849,56 @@ private void logExpressions(List<RuleSetBean> validRuleSets) {
      * @see org.akaza.openclinica.service.rule.RuleSetServiceInterface#filterRuleSetsByGroupOrdinal(java.util.List)
      */
     public List<RuleSetBean> filterRuleSetsByGroupOrdinal(List<RuleSetBean> ruleSets) {
-
-        for (RuleSetBean ruleSetBean : ruleSets) {
+  	         
+    	for (RuleSetBean ruleSetBean : ruleSets) {
             List<ExpressionBean> expressionsWithCorrectGroupOrdinal = new ArrayList<ExpressionBean>();
+            
+           /**
+	       	 * get the itemDataSize by studyEventId,itemGroupOid,itemOid for the whole study
+	       	 * only run once
+	       	 */  
+	       	String itemOid = getExpressionService().getItemOid(ruleSetBean.getExpressions().get(0).getValue());
+	        String itemGroupOid = getExpressionService().getItemGroupOid(ruleSetBean.getExpressions().get(0).getValue());
+           
+            HashMap itemDataCountHm = getItemDataDao().findCountByStudyEventAndOIDs(ruleSetBean.getStudyId(), itemOid, itemGroupOid);
+           
+            
+            int itemDataSize = 0;
+            
             for (ExpressionBean expression : ruleSetBean.getExpressions()) {
+          	  
                 String studyEventId = getExpressionService().getStudyEventDefinitionOrdninalCurated(expression.getValue());
-                String itemOid = getExpressionService().getItemOid(expression.getValue());
-                String itemGroupOid = getExpressionService().getItemGroupOid(expression.getValue());
+                itemOid = getExpressionService().getItemOid(expression.getValue());
+                itemGroupOid = getExpressionService().getItemGroupOid(expression.getValue());
                 String groupOrdinal = getExpressionService().getGroupOrdninalCurated(expression.getValue());
-                List<ItemDataBean> itemDatas = getItemDataDao().findByStudyEventAndOids(Integer.valueOf(studyEventId), itemOid, itemGroupOid);
-                logger.debug("studyEventId {} , itemOid {} , itemGroupOid {} , groupOrdinal {} , itemDatas {}", new Object[] { studyEventId, itemOid,
-                    itemGroupOid, groupOrdinal, itemDatas.size() });
-
+              
+               
+                String key = studyEventId + itemGroupOid +itemOid;
```

Comment:
instead of grouping the results and getting a count by this key, can you pass this key into the query itself? That way, we wouldn't have to get multiple counts.

Reply:
if pass this key into the query itself, then this means also need to put the query in the same loop level of where we get the key: studyEventId + itemGroupOid +itemOid, this is in loop:  for (ExpressionBean expression : ruleSetBean.getExpressions()) {...}, 
 so this refactor  will only just change the query to get the item data count, instead of item data objects, but the major performance issue is still there, the similar get count sql query will run 100K times in this attached test case.

Reply:
as discussed, calculate the item count per study event by moving the logic from the first for loop into the for loop below.

Reply:
done, please review